k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -0,0 +1,122 @@
+package io.buoyant.k8s
+
+import com.twitter.concurrent.AsyncStream
+import com.twitter.finagle.http
+import com.twitter.finagle.http.Response
+import com.twitter.finagle.param.HighResTimer
+import com.twitter.finagle.service.{Backoff, RetryBudget, RetryFilter, RetryPolicy}
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.io.Reader
+import com.twitter.util._
+import com.twitter.util.TimeConversions._
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * An abstract class that encapsulates the ability to Watch a k8s [[Resource]].
+ */
+private[k8s] abstract class Watchable[O <: KubeObject: Manifest, W <: Watch[O]: Manifest]
+  extends Resource {
+  protected def backoffs: Stream[Duration]
+  protected def path: String
+  protected def stats: StatsReceiver
+
+  /**
+   * implementing classes should define this method to retrieve the current state
+   * of the resource to pass as Modified results to any watchers.
+   *
+   * @return a Future containing a sequence of Watches and an optional String representing the current resourceVersion
+   */
+  protected def restartWatches(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None
+  ): Future[(Seq[W], Option[String])]
+
+  protected def infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+    RetryPolicy.backoff(backoffs) {
+      // We will assume 5xx are retryable, everything else is not for now
+      case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+      case (_, Throw(NonFatal(ex))) =>
+        log.error(s"retrying k8s endpoints request on error $ex")
+        true
+    },
+    HighResTimer.Default,
+    stats,
+    RetryBudget.Infinite
+  )
+
+  /**
+   * Watch this resource for changes, using a chunked HTTP request.
+   *
+   * TODO: investigate k8s websockets support as an alternative HTTP approach
+   *
+   * @return An asynchronous stream of [[Watch]] objects containing additions, modifications, and
+   *         deletions to the items in the list being watched, and a Closable handle allowing for
+   *         the watch to be terminated.
+   */
+  def watch(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None,
+    resourceVersion: Option[String] = None
+  ): (AsyncStream[W], Closable) = {
+    val close = new AtomicReference[Closable](Closable.nop)
+
+    // Internal method used to recursively retry watches as needed on failures.
```

Comment:
personal preference: it would be more readable to me if these internal methods were instead `private[this]` methods on `Watchable`